### PR TITLE
Metadata Handling

### PR DESCRIPTION
**Pull request by jeanggi90**
Related to #100. I do not think that there is one set of metadata which suit the need to all users. Neither do I think that all users want to fiddle with the source code in order to get their desired metadata to show. Therefore I have added an option to set the desired metadata as a list in the settings.
Since the data is treated generically, I have not yet figured out the best was to properly format the data, e.g. add units, as `piexif` does not seem to have a notion of the units themselves.

Furthermore, depending on the stage of my photography workflow, different metadata information are relevant. Therefore I have thought about adding 3 to 5 user customizable sets of metadata key, which can be viewed one the metadata widget. Changing the metadata set should be done using a shortcut (e.g. <kbd>CTRL</kbd> + <kbd>1</kbd> through <kbd>CTRL</kbd> + <kbd>5</kbd>).

@karlch tell me what you think about that

### CODE REVIEW

**Comment by karlch** [comment]
Thanks for this PR, I like it a lot!

I am also not quite sure what to do concerning the units. To be fair I am also fine with just leaving them out. If you think this is not an option, we could keep our own dictionary mapping exif key names to their corresponding units, falling back to no unit. So something along:
```
units = {
    "ExposureTime": "s",
}
```
Or even with the corresponding structure from piexif:
```
units = {
    piexif.ExifIFD: {
        piexif.ExifIFD.ExposureTime: "s",
    }
}
```
and them add the units on-demand / when we notice something is missing. I am not sure the maintenance overhead is worth it though. As much as I should run aways from numbers without units as a physicist :smile: 

Now to your second point, the user-configurable sets of metadata keys. I also find this a brilliant idea, but would recommend to put this into a second PR after the current one to keep it simple as this is probably quite a bit more effort. If you don't mind, I would open a new issue for this where I could give some basic hints on a possible implementation. You could then take over if you like.

If it is not too much to ask, I would leave a few comments concerning the PR code that could be addressed before merging. It would be great if you could also fix the issues raised by the `lint` ci. Most of them should be gone once [running pre-commit over the code](https://karlch.github.io/vimiv-qt/documentation/hacking#style-and-formatting). Don't worry if you don't get to it, I can also fix this after merging.

Thanks again, much appreciated!

**Comment by jeanggi90** [comment]
I am glad that you like my idea as well as implementation. I must say it is a real pleasure to work with your clean, well structured and well documented code!

The unit is actually not the only problem, but the formatting of the data is also a not right. For example the shutterspeed is normally given as a fraction like `1/250 s` whereas the f-stop is given in floats like `f/5.6`. In my view, this formatting (including) the units is something which should be provided by the metadata reader tool. For example the [exiv2](https://www.exiv2.org/) does a good job at that. However, in the [python wrapper](https://python3-exiv2.readthedocs.io/en/latest/index.html) I did not find this option. 
Anyways, if you are fine with how it is at the moment, I would suggest that we leave it as it is, but open and issue which addresses the formatting and the units of the metadata.

I think it makes sense to open a second PR for my other idea. Especially, since I am not reluctant in getting additional information and hints concerning the implementation (esp. how to deal with the different information displayed in the metadata widget and how to add new commands/keybindings etc.) :blush: 

Back to the current PR. If you agree, I will do the final cleanup, fix the linting error and then remove the draft flag.

File: vimiv/api/settings.py [hunk 1]
```diff
@@ -488,3 +488,14 @@ class title:  # pylint: disable=invalid-name
         desc="Default window title if no mode specific options exist",
     )
     StrSetting("title.image", "vimiv - {basename}", desc="Window title in image mode")
+
+
+class metadata:  # pylint: disable=invalid-name
+    """Namespace for metadata related settings."""
+
+    # Store the keys as a comma seperated string
+    StrSetting(
```
Comment (karlch) [inline vimiv/api/settings.py:497]: It would be great to asign the setting to a variable, e.g.: `keylist = StrSetting( ...`. Like this we can refer to it later via the python object, not only as a string.

File: vimiv/imutils/exif.py [hunk 2]
```diff
@@ -99,40 +100,42 @@ def _load_exif(self, filename):
...
+            for tag in self._exif[ifd]:
+                keyname = piexif.TAGS[ifd][tag]["name"]
+                keytype = piexif.TAGS[ifd][tag]["type"]
+                val = self._exif[ifd][tag]
+                _logger.debug(
+                    f"name: {keyname} type: {keytype} value: {val} tag: {tag}"
+                )
+                if keyname.lower() not in desired_keys:
+                    _logger.debug(
+                        f"{keyname.lower()} not in {desired_keys}. Ignoring it"
```
Comment (karlch) [inline vimiv/imutils/exif.py:124]: Maybe another reference to `{desired_keys}` is a bit too verbose. I would personally prefer to just have "desired keys" as a string, but leaving it as it is is fine too.

File: vimiv/imutils/exif.py [hunk 3]
```diff
@@ -99,40 +100,42 @@ def _load_exif(self, filename):
...
+                    )
+                    continue
+                if keytype in (1, 3, 4, 9):  # integer
+                    with contextlib.suppress(KeyError):
```
Comment (karlch) [inline vimiv/imutils/exif.py:128]: As far as I can see all of these calls to `contextlib.suppress` are not needed as we are just adding a value to the dict, not retrieving anything. Thus `KeyError` should never be raised, or am I missing something?

File: vimiv/imutils/exif.py [hunk 4]
```diff
@@ -99,40 +100,42 @@ def _load_exif(self, filename):
...
+                val = self._exif[ifd][tag]
+                _logger.debug(
+                    f"name: {keyname} type: {keytype} value: {val} tag: {tag}"
+                )
+                if keyname.lower() not in desired_keys:
+                    _logger.debug(
+                        f"{keyname.lower()} not in {desired_keys}. Ignoring it"
+                    )
+                    continue
+                if keytype in (1, 3, 4, 9):  # integer
```
Comment (karlch) [inline vimiv/imutils/exif.py:127]: Concerning the keytype, are you sure concerning these numbers? Would it be possible to use the values from `piexif.TYPES` instead, so we don't rely on the integer values? So something like:
```
if keytype in (piexif.TYPES.Short, piexif.TYPES.Long, piexif.TYPES.SShort, piexif.TYPES.SLong):
    ....
```

File: vimiv/imutils/exif.py [hunk 5]
```diff
@@ -99,40 +100,42 @@ def _load_exif(self, filename):
...
+                        self[keyname] = val
+                elif keytype == 2:  # byte encoded ascii
+                    with contextlib.suppress(KeyError):
+                        self[keyname] = val.decode()
+                elif keytype in (5, 10):  # (int, int) <=> numerator, denominator
+                    with contextlib.suppress(KeyError):
+                        self[keyname] = f"{val[0]}/{val[1]}"
+                elif keytype == 7:  # byte
+                    with contextlib.suppress(KeyError):
+                        self[keyname] = val.decode()
```
Comment (karlch) [inline vimiv/imutils/exif.py:138]: This is the same logic as above for the byte encoded ascii, so maybe they could be merged with a `keytype in` like for the other values?

File: vimiv/imutils/exif.py [hunk 6]
```diff
@@ -99,40 +100,42 @@ def _load_exif(self, filename):
         """Load exif information from filename into the dictionary."""
         try:
             self._exif = piexif.load(filename)
+            desired_keys = [
+                e.lower().strip()
+                for e in api.settings.get_value("metadata.keylist").split(",")
```
Comment (karlch) [inline vimiv/imutils/exif.py:105]: If the setting was stored in a variable, you could refer to it here via the object. I.e.:
```
for e in api.settings.metadata.keylist.value.split(", ")
```
I admit this is quite a list of `.` chained stuff, but at least we don't rely on the string.

File: vimiv/api/settings.py (hunk 1 above)
Comment (jeanggi90) [reply vimiv/api/settings.py:497]: Surely, I have noticed that you sometimes assign the setting to a variable and sometimes you dont. May I ask you want the advantage or point of doing in one or the other way?

File: vimiv/imutils/exif.py (hunk 4 above)
Comment (jeanggi90) [reply vimiv/imutils/exif.py:127]: You are right. I did not think about that. Thanks for pointing that out!

File: vimiv/imutils/exif.py (hunk 6 above)
Comment (jeanggi90) [reply vimiv/imutils/exif.py:105]: I see, where it comes handy to store the setting in a variable. :blush:  However, I would suggest to add an`.strip()`.This way it is more robust when do user does not get the formatting fully right

File: vimiv/api/settings.py (hunk 1 above)
Comment (karlch) [reply vimiv/api/settings.py]: The general rule is to use the variable where possible as we can then make use of actual python objects directly. This also gives us access to the `updated` signal which allows objects to react when a setting changed, e.g. the slideshow in `imutils.slideshow` has a timer that updates its interval upon `api.settings.slideshow.delay.changed`.

Of course there are places where we cannot use the type immediately, for example when running the `:set` command on the command-line, we first retrieve the object by its name as only the name can be passed as a string by the user.

If there are only cases where the setting is used by name and it cannot be used by its type, the setting is not assigned to al variable. This is currently only the case for status-related settings, for example `statusbar.left` and `statusbar.left_image`. Here we can define arbitrary combinations of `position_mode`, but not all have defaults. We also only retrieve the value via this string pattern as we can then iterate over the positions and make use of the current mode name. This happens for example in `gui.statusbar` in the `_get_text(position, mode)` method.

tl;dr assign where possible, use string where the code is too dynamic to access the variable directly

Hope this makes it a bit more clear instead of more confusing :smile:

**Comment by karlch** [comment]
Thanks again! :tada:
